File: adapters/java/src/main/java/adapter/Float16.java
package adapter;

/**
 * Hand-rolled IEEE-754 binary16 (f16) &lt;-&gt; binary64 (f64) bit-level
 * conversion. No dependency-free JVM library performs this; ported from the
 * same round-to-nearest-even algorithm the Rust adapter's `half` crate uses
 * (mirrors adapters/kotlin/src/main/kotlin/Float16.kt bit-for-bit).
 */
public final class Float16 {
    private Float16() {}

    public static int doubleToF16Bits(double value) {
        long x = Double.doubleToRawLongBits(value);
        long sign = (x >>> 63) & 1L;
        long exp = (x >>> 52) & 0x7ffL;
        long man = x & 0xfffffffffffffL;

        if (exp == 0x7ffL) {
            long newMan = man >>> 42;
            long nanBit = man != 0L ? 1L << 9 : 0L;
            return (int) (((sign << 15) | 0x7c00L | newMan | nanBit) & 0xffffL);
        }

        long halfSignBits = sign << 15;
        long halfExp = exp - 1023L + 15L;

        if (halfExp >= 0x1fL) {
            return (int) ((halfSignBits | 0x7c00L) & 0xffffL);
        }

        if (halfExp <= 0L) {
            if (10L - halfExp > 21L) {
                return (int) (halfSignBits & 0xffffL);
            }
            long manWithHidden = man | 0x10000000000000L;
            long halfMan = manWithHidden >>> (int) (43L - halfExp);
            long roundBit = 1L << (int) (42L - halfExp);
            if ((manWithHidden & roundBit) != 0L && (manWithHidden & (3L * roundBit - 1L)) != 0L) {
                halfMan += 1L;
            }
            return (int) ((halfSignBits | halfMan) & 0xffffL);
        }

        long halfExpBits = halfExp << 10;
        long halfMan = man >>> 42;
        long roundBit = 1L << 41;
        long combined = halfSignBits | halfExpBits | halfMan;
        boolean roundUp = (man & roundBit) != 0L && (man & (3L * roundBit - 1L)) != 0L;
        return (int) ((roundUp ? combined + 1L : combined) & 0xffffL);
    }

    public static double f16BitsToDouble(int bits) {
        long i = bits & 0xffffL;
        long sign = (i >>> 15) & 1L;
        long exp = (i >>> 10) & 0x1fL;
        long man = i & 0x3ffL;
        long sign64 = sign << 63;

        if (exp == 0L && man == 0L) {
            return Double.longBitsToDouble(sign64);
        }
        if (exp == 0L) {
            // Normalize the 10-bit subnormal significand by left-shifting
            // until its leading set bit reaches position 10 (the
            // implicit-bit position), counting shifts via expAdj. expAdj
            // MUST start at 1, not -1: for the smallest subnormal (man=1,
            // true value 2^-24), exactly 10 shifts are needed to reach
            // manAdj=0x400, and the final double exponent must come out to
            // 1023-24=999. That only holds if expAdj lands on 1-10=-9
            // (giving 1008+(-9)=999); a -1 starting point yields -11,
            // undercounting by 2 and silently quartering the reconstructed
            // magnitude of every subnormal f16 value.
            long expAdj = 1L;
            long manAdj = man;
            while ((manAdj & 0x400L) == 0L) {
                manAdj <<= 1;
                expAdj -= 1L;
            }
            manAdj &= 0x3ffL;
            long exp64 = 1023L - 15L + expAdj;
            long man64 = manAdj << 42;
            return Double.longBitsToDouble(sign64 | (exp64 << 52) | man64);
        }
        if (exp == 0x1fL) {
            if (man == 0L) {
                return Double.longBitsToDouble(sign64 | 0x7ff0000000000000L);
            }
            return Double.longBitsToDouble(sign64 | 0x7ff0000000000000L | (man << 42));
        }
        long exp64 = exp - 15L + 1023L;
        long man64 = man << 42;
        return Double.longBitsToDouble(sign64 | (exp64 << 52) | man64);
    }

    private static boolean sameBits(double a, double b) {
        return Double.doubleToRawLongBits(a) == Double.doubleToRawLongBits(b);
    }

    /** Callers must handle NaN before reaching these helpers -- sameBits's
     * raw-bits comparison never matches NaN to NaN. */
    public static byte[] tryF16(double f) {
        int bits = doubleToF16Bits(f);
        double back = f16BitsToDouble(bits);
        if (!sameBits(back, f)) return null;
        return new byte[] { (byte) ((bits >>> 8) & 0xff), (byte) (bits & 0xff) };
    }

    public static byte[] tryF32(double f) {
        float v = (float) f;
        double back = v;
        if (!sameBits(back, f)) return null;
        int bits = Float.floatToRawIntBits(v);
        return new byte[] {
            (byte) ((bits >>> 24) & 0xff),
            (byte) ((bits >>> 16) & 0xff),
            (byte) ((bits >>> 8) & 0xff),
            (byte) (bits & 0xff),
        };
    }
}

File: adapters/c/src/float16.c
#include "float16.h"

#include <string.h>

static uint64_t double_bits(double v) {
    uint64_t u;
    memcpy(&u, &v, sizeof(u));
    return u;
}

static double bits_double(uint64_t u) {
    double v;
    memcpy(&v, &u, sizeof(v));
    return v;
}

static uint32_t float_bits(float v) {
    uint32_t u;
    memcpy(&u, &v, sizeof(u));
    return u;
}

uint16_t double_to_f16_bits(double value) {
    uint64_t x = double_bits(value);
    uint64_t sign = (x >> 63) & 1;
    uint64_t exp = (x >> 52) & 0x7ff;
    uint64_t man = x & 0xfffffffffffffULL;

    if (exp == 0x7ff) {
        /* Infinity or NaN. */
        uint64_t new_man = man >> 42;
        uint64_t nan_bit = man != 0 ? (1ULL << 9) : 0;
        return (uint16_t)((sign << 15 | 0x7c00 | new_man | nan_bit) & 0xffff);
    }

    uint64_t half_sign_bits = sign << 15;
    int64_t half_exp = (int64_t)exp - 1023 + 15;

    if (half_exp >= 0x1f) {
        return (uint16_t)((half_sign_bits | 0x7c00) & 0xffff);
    }

    if (half_exp <= 0) {
        if (10 - half_exp > 21) {
            /* Underflows even a subnormal half: signed zero. */
            return (uint16_t)(half_sign_bits & 0xffff);
        }
        uint64_t man_with_hidden = man | 0x10000000000000ULL;
        uint64_t half_man = man_with_hidden >> (uint64_t)(43 - half_exp);
        uint64_t round_bit = ((uint64_t)1) << (uint64_t)(42 - half_exp);
        if ((man_with_hidden & round_bit) != 0 && (man_with_hidden & (3 * round_bit - 1)) != 0) {
            half_man += 1;
        }
        return (uint16_t)((half_sign_bits | half_man) & 0xffff);
    }

    uint64_t half_exp_bits = ((uint64_t)half_exp) << 10;
    uint64_t half_man = man >> 42;
    uint64_t round_bit = ((uint64_t)1) << 41;
    uint64_t combined = half_sign_bits | half_exp_bits | half_man;
    bool round_up = (man & round_bit) != 0 && (man & (3 * round_bit - 1)) != 0;
    if (round_up) combined += 1;
    return (uint16_t)(combined & 0xffff);
}

double f16_bits_to_double(uint16_t bits) {
    uint64_t i = bits;
    uint64_t sign = (i >> 15) & 1;
    uint64_t exp = (i >> 10) & 0x1f;
    uint64_t man = i & 0x3ff;
    uint64_t sign64 = sign << 63;

    if (exp == 0 && man == 0) {
        return bits_double(sign64);
    }
    if (exp == 0) {
        /* Normalize the 10-bit subnormal significand by left-shifting until
         * its leading set bit reaches position 10 (the implicit-bit
         * position), counting shifts via exp_adj. exp_adj MUST start at 1,
         * not -1: for the smallest subnormal (man=1, true value 2^-24),
         * exactly 10 shifts are needed to reach manAdj=0x400, and the
         * final double exponent must come out to 1023-24=999. That only
         * holds if exp_adj lands on 1-10=-9 (giving 1008+(-9)=999); a
         * -1 starting point yields -11, undercounting by 2 and silently
         * halving (well, quartering: 2^-2) the reconstructed magnitude of
         * every subnormal f16 value. Verified against a from-scratch
         * derivation and an exhaustive f16-bit-pattern round-trip test
         * (tests/test_float16.c) covering all 2046 subnormal patterns. */
        int64_t exp_adj = 1;
        uint64_t man_adj = man;
        while ((man_adj & 0x400) == 0) {
            man_adj <<= 1;
            exp_adj -= 1;
        }
        man_adj &= 0x3ff;
        int64_t exp64 = 1023 - 15 + exp_adj;
        uint64_t man64 = man_adj << 42;
        return bits_double(sign64 | ((uint64_t)exp64 << 52) | man64);
    }
    if (exp == 0x1f) {
        if (man == 0) {
            return bits_double(sign64 | 0x7ff0000000000000ULL);
        }
        return bits_double(sign64 | 0x7ff0000000000000ULL | (man << 42));
    }
    uint64_t exp64 = exp - 15 + 1023;
    uint64_t man64 = man << 42;
    return bits_double(sign64 | (exp64 << 52) | man64);
}

bool same_bits(double a, double b) {
    return double_bits(a) == double_bits(b);
}

bool try_f16(double f, uint8_t out[2]) {
    uint16_t bits = double_to_f16_bits(f);
    double back = f16_bits_to_double(bits);
    if (!same_bits(back, f)) return false;
    out[0] = (uint8_t)(bits >> 8);
    out[1] = (uint8_t)(bits & 0xff);
    return true;
}

bool try_f32(double f, uint8_t out[4]) {
    float v = (float)f;
    double back = (double)v;
    if (!same_bits(back, f)) return false;
    uint32_t bits = float_bits(v);
    out[0] = (uint8_t)(bits >> 24);
    out[1] = (uint8_t)(bits >> 16);
    out[2] = (uint8_t)(bits >> 8);
    out[3] = (uint8_t)(bits & 0xff);
    return true;
}
